circus-train-aws-sns/src/main/java/com/hotels/bdp/circustrain/aws/sns/event/SnsListener.java

```diff
@@ -118,29 +118,44 @@ public void tableReplicationStart(EventTableReplication tableReplication, String
 
   @Override
   public void tableReplicationSuccess(EventTableReplication tableReplication, String eventId) {
-    String endTime = clock.getTime();
-    EventReplicaTable replicaTable = tableReplication.getReplicaTable();
-    SnsMessage message = new SnsMessage(SnsMessageType.SUCCESS, config.getHeaders(), startTime, endTime, eventId,
-        sourceCatalog.getName(), replicaCatalog.getName(), replicaCatalog.getHiveMetastoreUris(),
-        tableReplication.getSourceTable().getQualifiedName(), tableReplication.getQualifiedReplicaName(),
-        replicaTable.getTableLocation(), partitionKeyTypes,
-        getModifiedPartitions(partitionsToAlter, partitionsToCreate), getBytesReplicated(), null);
-    publish(config.getSuccessTopic(), message);
+    try {
+      String endTime = clock.getTime();
+      EventReplicaTable replicaTable = tableReplication.getReplicaTable();
+      SnsMessage message = new SnsMessage(SnsMessageType.SUCCESS, config.getHeaders(), startTime, endTime, eventId,
+          sourceCatalog.getName(), replicaCatalog.getName(), replicaCatalog.getHiveMetastoreUris(),
+          tableReplication.getSourceTable().getQualifiedName(), tableReplication.getQualifiedReplicaName(),
+          replicaTable.getTableLocation(), partitionKeyTypes,
+          getModifiedPartitions(partitionsToAlter, partitionsToCreate), getBytesReplicated(), null);
+      publish(config.getSuccessTopic(), message);
+    } finally {
+      // reset registered partitions between database cycles
+      partitionsToCreate = null;
+      partitionsToAlter = null;
+      partitionKeyTypes = null;
+    }
   }
 
   @Override
   public void tableReplicationFailure(EventTableReplication tableReplication, String eventId, Throwable t) {
-    if (startTime == null) {
-      startTime = clock.getTime();
+    try {
+      if (startTime == null) {
```

Comment:
This code was already here but I wonder why we have this null check and change to `startTime` in `tableReplicationFailure` but not in `tableReplicationSuccess`? This makes me think though - should we also set `startTime` to null at the end of both of these methods?

Reply:
Maybe because of the functions are used in `Locomotive.java`:
```java
try {
  Replication replication = replicationFactory.newInstance(tableReplication);
  tableReplicationListener.tableReplicationStart(...);
  tableReplicationListener.tableReplicationSuccess(...);
} catch (Throwable t) {
  replicationFailures++;
  completionCode = CompletionCode.FAILURE;
  LOG.error("Failed to replicate: {}.", summary, t);
  tableReplicationListener.tableReplicationFailure(...);
```

If there's an issue with creating a new instance, then `startTime` might not have been initialised, but if everything works fine, `tableReplicationSuccess` should have the startTime that was set in `tableReplicationStart`.

So maybe this is fine?

Reply:
Yeah, that makes sense. There's quite a lot of assumptions made in this class based on the order that things are called on it which isn't ideal but that's a different topic ;)

When the above failure happens is the listener ever called again? i.e. could the time get set in the failure, and it then tries to do something else which also fails, at which point the time in the failure would be the previous failure's start time?

Reply:
Yes, I'm fine with leaving the assumptions on call order for another day. I just wonder if it's worth clearing the startTime though?

Reply:
that's true, that's but in that case the whole start-end time is rather useless info anyway.
Pretty sure the null check is there to have a quick nullPointerException fix and have "some" value in the failure message.

Reply:
I mean by all means set startTime to null after a success, just in case. That won't hurt.